src/main/java/org/truffleruby/core/format/read/array/ReadStringNode.java

```diff
@@ -64,6 +70,11 @@ private Object readAndConvert(VirtualFrame frame, Object value) {
                     WriteByteNodeGen.create(new LiteralFormatNode((byte) 0))));
         }
 
+//        TODO: wrap into a Node and call execute() on it instead of having the logic here?
```

Comment:
Yes, it would be better to wrap.
Nodes must be adopted, which means they need to be a `@Child` field of some other nodes, and wrapping `ReadStringNode` with a `TruncateStringNode` seems a good way to do that.

Reply:
Ah, right but as you mentioned above, we need to do it before ` toStringNode.executeToString` otherwise we lose the encoding, etc and are left only with a `byte[]`.
So then we can reverse it, so like you do it here, but add a `@Child TruncateStringNode truncateStringNode` field in this class, assign it in the constructor if precision isn't `null`, and use it here.
You'll want a separate execute method like `execute(RubyString)`.
The regular `execute(VirtualFrame)` can just `throw CompilerDirectives.shouldNotReachHere()`, it's not going to be used.